tools/cldr-code/src/main/java/org/unicode/cldr/util/personname/PersonNameFormatter.java

```diff
@@ -581,6 +681,86 @@ public boolean equals(Object obj) {
+            // if 3 of the 4 rows are equal, merge the other two
+            if (lengths.equals(that.lengths)
+                && styles.equals(that.styles)
+                && usages.equals(that.usages)) {
+                return new ParameterMatcher(lengths, styles, usages, Sets.union(orders, that.orders));
+            } else if (lengths.equals(that.lengths)
+                && styles.equals(that.styles)
+                && orders.equals(that.orders)) {
+                return new ParameterMatcher(lengths, styles, Sets.union(usages, that.usages), orders);
+            } else if (lengths.equals(that.lengths)
+                && usages.equals(that.usages)
+                && orders.equals(that.orders)) {
+                return new ParameterMatcher(lengths, Sets.union(styles, that.styles), usages, orders);
+            } else if (styles.equals(that.styles)
+                && usages.equals(that.usages)
+                && orders.equals(that.orders)) {
+                return new ParameterMatcher(Sets.union(lengths, that.lengths), styles, usages, orders);
+            }
+            return null;
+        }
+
+        public ParameterMatcher slim() {
+            return new ParameterMatcher(
+                lengths.equals(Length.ALL) ? ImmutableSet.of() : lengths,
+                    styles.equals(Style.ALL) ? ImmutableSet.of() : styles,
+                        usages.equals(Usage.ALL) ? ImmutableSet.of() : usages,
+                            orders.equals(Order.ALL) ? ImmutableSet.of() : orders
+                );
+        }
+
+        /**
+         * Compact a set of FormatParameters into a (smaller) collection of ParameterMatchers
+         */
+        public static Set<ParameterMatcher> compact(Collection<FormatParameters> expanded) {
+            Set<ParameterMatcher> result = new TreeSet<>();
+            for (FormatParameters item : expanded) {
+                result.add(new ParameterMatcher(item));
+            }
+            // try merging each pair
+            // if we can merge, then start over from the top
+            // look at optimizing later
+            main:
+                while (true) {
+                    for (ParameterMatcher item1 : result) {
+                        for (ParameterMatcher item2: result) {
+                            if (item1 == item2) { // skip ourselves
+                                continue;
+                            }
+                            ParameterMatcher item12 = item1.merge(item2); // merge if possible
+                            if (item12 != null) {
+                                result.remove(item1);
+                                result.remove(item2);
+                                result.add(item12);
+                                continue main; // retry everything
+                            }
+                        }
+                    }
+                    break;
+                }
+
+            // now replace any "complete" items by empty.
```

Comment:
Does it really need to happen? IMO, if you to keep attribute sets non-empty, this will simplify logic of the matcher: no need to treat "empty set" in a special way, just check if the item is in the set.

The slimming would probably make sense if you used this logic to generate XML or something, but if it's used only as an internal structure, I'd rather not do that.

Reply:
I went ahead and merged this, so that you can go ahead and rebase on it. I'll look at your comments in a subsequent PR.